### DISCUSSION TITLE
Authentication for Admin Panel

### DISCUSSION THREAD

**Discussion by MohammedHafiz** [supabase/supabase#5657] [Questions] [answered]
Hey,
I have made an app using supabase and now I'm planning to build an admin panel with a different front end for it. What authentication should I use for this admin panel? Are there any particular login credentials available?

Mohammed Hafiz BA

**Comment by awalias** [comment, marked as answer]
you should probably use role based access control - there is an example of admin user roles in this README: 

https://github.com/supabase/supabase/tree/master/examples/nextjs-slack-clone#role-based-access-control-rbac

[Example schema here](https://github.com/supabase/supabase/tree/master/examples/nextjs-slack-clone#postgres-row-level-security)

**Reply by MohammedHafiz** [reply]
Thanks for the reply however this is a little complicated for my current experience. Could you just tell me how I can map an email from auth.users to bypass RLS policy based on id? For example, I have a policy on a table where users can read only their own profiles (uid() = id). Can I add an OR condition like (uid() = id || uid() = * my-admin’s-uuid*) ?

**Reply by GaryAustin1** [reply]
RLS is just SQL.  You can use any SQL including selects, function calls, etc. as long as final result is true or false
So yes you can do or conditions.
You could  do something like (uid() = id || auth.email() = 'cat@cats.com')    

Here are some Supabase examples
https://supabase.com/docs/guides/auth/row-level-security

**Reply by MohammedHafiz** [reply]
I tried auth.email() = 'my-admin-email' but it still doesn't return the entire row. How do I achieve this?

**Reply by GaryAustin1** [reply]
Can you show your policy that works for a user to see their own row and the entire policy for the admin.  I assume by entire row you mean all the rows.

**Reply by MohammedHafiz** [reply]
So this is a completed orders table where a user can select all the rows based on their ID

(uid() = user_id) is the current policy
I'd like to return all the rows(complete table) for one particular user

**Reply by GaryAustin1** [reply]
Yes, hence adding the or condition with that users email to the policy.  You said you tried it, auth.email() should work.  I'm not sure on casing of email in jwt which is what email() uses.

**Reply by MohammedHafiz** [reply]
OKay so when I tried updating (uid() = user_id || auth.email() = "myadminemail") I could not save the policy. I got a syntax error near "=". Instead I added another policy for select, auth.email() = "myadminemail" and it still ended up following the first policy that I had added. Didn't return all the rows.

**Reply by burggraf** [reply]
Did you mean to use `or` here instead of `||`?  This is sql so should be using sql syntax and not javascript syntax.

**Reply by burggraf** [reply]
Also double quotes are not appropriate here, in SQL we need single quotes:

`(uid() = user_id OR auth.email() = 'cats@cats.com')`

**Reply by burggraf** [reply]
and if `uid()` doesn't work you may need to call `auth.uid()`

**Reply by GaryAustin1** [reply]
Crap totally missed the using || instead of OR in copying his example...

**Reply by burggraf** [reply]
I do it all the time (js vs sql).  LOL

**Comment by MohammedHafiz** [comment]
Thanks for the reply @burggraf @GaryAustin1 . It worked for me now.

**Comment by bennik88** [comment]
The Links in the answer are sadly no longer available would someone be able to renew them?

**Reply by GaryAustin1** [reply]
So I just looked up one and you can see Supabase has moved things with the js version 1 versus 2 out.
I suspect alot of github links will be broken....
https://github.com/supabase/examples/tree/main/supabase-js-v1/slack-clone/nextjs-slack-clone#role-based-access-control-rbac
https://github.com/supabase/supabase/tree/master/examples/nextjs-slack-clone#role-based-access-control-rbac

**Reply by bennik88** [reply]
Thank you a lot for your answer. I worked myself through all your documentation and RLS Videos on Youtube on your Channel and searched a lot over google but I'm absolutely finding no content on how to work with different roles like "admin", "superadmin", "client" or whatever custom roles I want. I'm coming from Backends like Django, Firebase and MongoDB and now read my way through all your Triggers, Serverless functions, and multiple SQL Curses but can't get it to work like Firebase with Custom Claims or Django Role Based Authentication. You're providing an abstract UI for Users who don't want to worry about always diving into PosergreSQL but I'm not able to figure out how to deal with such a basic use case when it comes to authentication. I've also seen your SlackClone but I don't want to work with email address RLS I simply want a role like "superadmin" which is stored at the user or a related table who is then able to access (create, update, delete) all resources and also set roles for authenticated users. How would I achieve this? I also didn't find any information on how to change the user.role() attribute or extend it. Did I miss something in the docs as this is such a basic usecase? Restricting access for "signed in / authenticated" users with RLS is clear for me. I'm replying here because it's the first tread on google when it comes to this topic, so therefore other users hopefully find it as well.

**Reply by TorMagne** [reply]
I will just leave a comment here and say that i agree with @bennik88 on this. I would have been nice of they had an UI like Strapi where you can set different roles and give user admin, or moderator etc
